shell/common/rasterizer.cc

```diff
@@ -125,51 +124,89 @@ void Rasterizer::DisableThreadMergerIfNeeded() {
 }
 
 void Rasterizer::NotifyLowMemoryWarning() const {
-  if (!surface_) {
+  if (!studio_) {
     FML_DLOG(INFO)
         << "Rasterizer::NotifyLowMemoryWarning called with no surface.";
     return;
   }
-  auto context = surface_->GetContext();
+  auto context = studio_->GetContext();
   if (!context) {
     FML_DLOG(INFO)
         << "Rasterizer::NotifyLowMemoryWarning called with no GrContext.";
     return;
   }
-  auto context_switch = surface_->MakeRenderContextCurrent();
+  auto context_switch = studio_->MakeRenderContextCurrent();
   if (!context_switch->GetResult()) {
     return;
   }
   context->performDeferredCleanup(std::chrono::milliseconds(0));
 }
 
+void Rasterizer::AddSurface(int64_t view_id, std::unique_ptr<Surface> surface) {
+  bool insertion_happened =
+      surfaces_
+          .try_emplace(/* map key=*/view_id, /*constructor args:*/ view_id,
+                       std::move(surface))
+          .second;
+  if (!insertion_happened) {
+    FML_DLOG(INFO) << "Rasterizer::AddSurface called with an existing view ID "
+                   << view_id << ".";
+  }
+}
+
 std::shared_ptr<flutter::TextureRegistry> Rasterizer::GetTextureRegistry() {
   return compositor_context_->texture_registry();
 }
 
 GrDirectContext* Rasterizer::GetGrContext() {
-  return surface_ ? surface_->GetContext() : nullptr;
+  return studio_ ? studio_->GetContext() : nullptr;
 }
 
-flutter::LayerTree* Rasterizer::GetLastLayerTree() {
-  return last_layer_tree_.get();
+bool Rasterizer::HasLastLayerTree() const {
+  // TODO(dkwingsmt): This method is only available in unittests now
+  for (auto& record_pair : surfaces_) {
+    auto& layer_tree = record_pair.second.last_tree;
+    if (layer_tree) {
+      return true;
+    }
+  }
+  return false;
 }
 
-void Rasterizer::DrawLastLayerTree(
-    std::unique_ptr<FrameTimingsRecorder> frame_timings_recorder) {
-  if (!last_layer_tree_ || !surface_) {
-    return;
+int Rasterizer::DrawLastLayerTree(
+    std::unique_ptr<FrameTimingsRecorder> frame_timings_recorder,
+    bool enable_leaf_layer_tracing) {
+  if (!studio_) {
+    return 0;
+  }
+  int success_count = 0;
+  bool should_resubmit_frame = false;
+  for (auto& record_pair : surfaces_) {
```

Comment:
Should this be able to render the last layer tree for just a single surface?

From my understanding, `DrawLastLayerTree` is used in scenarios where an external texture was updated but the framework's layer tree is otherwise unchanged. For example, if a video player needs to update the surface on each video frame even though the Flutter app itself hasn't rendered a frame. We shouldn't re-draw a second window that doesn't contain the video.

Reply:
If there are two videos in the app on two different windows, then we'll need the last-frame information for both views.

Reply:
Right, but do we need to rasterize all windows if only one window has a video player?

Reply:
I think you have a valid point, but let me rephrase your question.

In the case that multiple views have a video player each, we should probably allow `DrawLastLayerTree` to only draw the views that need a resubmission, i.e. whose video player triggered this vsync - instead of resubmitting all views on all vsyncs.

However, this feature is tricky. We need to allow VSyncs to carry information on the views that triggered it, so that the animator can send it back to the rasterizer.

Also, this feature can be done as a separate issue after the main changes, since it affects performance, not functionality.